Update distributed tracing docs

noahfalk opened the pull request:
We are about to announce support for OpenTelemetry .NET and there is
likely to be an influx of developers reading these docs. They are
already the #1 link for "distributed tracing" on bing.com so they are
probably getting above average traffic/scrutiny too.

- Reorganized the content into a landing page, conceptual docs,
instrumentation walkthroughs and collection walkthroughs
- Added some more intro and conceptual information assuming that many
developers are completely new to distributed tracing and have no
frame of reference.
- Removed the example of doing distributed tracing with DiagnosticSource
We might want to restore this in the future but hopefully it represents
a niche usage scenario that we are migrating away from. Without context
indicating when to use it it is likely to confuse new developers about
recommended patterns.
- Updated the examples to ensure they start with copy-pastable code
that is fully runnable.
- Added links for collecting distributed tracing instrumentation with
Application Insights

There are many oportunities for further improvement that we may
want to triage:
- Add examples and pictures of more realistic collected distributed
trace information to help developers internalize why it might be useful
to them.
- Add a diagnostic guide showing how distributed tracing assists in
resolving a realistic production issue.
- Ensure the getting started guides for collection transfer smoothly to
remote pages in the AppInsights and OpenTelemetry docs
  - There was a spot I wanted to link to supported OpenTelemetry
     exporters but there is no anchor that goes solely to that
     information.
- Flesh out the instrumentation guidance to answer more questions
developers may encounter such as:
  - naming conventions for Activity
  - when to use DisplayName
  - how to log exceptions
  - how to use links for batch processing systems
  - how to propagate IDs to and from network protocols
  - how to migrate pre-existing Activity instrumentation to use
     ActivitySource
- Add performance analysis and guidance to reassure developers that
  Activity instrumentation + OpenTelemetry are serious about high
  performance workloads.
- Discuss more conceptual topics and/or go into greater depth about
some areas:
  - Hierarchical vs. W3C ID.
  - Format of a W3C ID
  - Describe sampling options in ActivityListener
  - Explain why some activities are uncapturable with ActivityListener
     and require workarounds with DiagnosticListener.
 - Roadmap for handling issues around custom propagators, supporting
protocol updates, and suppressing undesired automatic Activity creation

noahfalk committed:
Update distributed tracing docs

We are about to announce support for OpenTelemetry .NET and there is
likely to be an influx of developers reading these docs. They are
already the #1 link for "distributed tracing" on bing.com so they are
probably getting above average traffic/scrutiny too.

- Reorganized the content into a landing page, conceptual docs,
instrumentation walkthroughs and collection walkthroughs
- Added some more intro and conceptual information assuming that many
developers are completely new to distributed tracing and have no
frame of reference.
- Removed the example of doing distributed tracing with DiagnosticSource
We might want to restore this in the future but hopefully it represents
a niche usage scenario that we are migrating away from. Without context
indicating when to use it it is likely to confuse new developers about
recommended patterns.
- Updated the examples to ensure they start with copy-pastable code
that is fully runnable.
- Added links for collecting distributed tracing instrumentation with
Application Insights

There are many oportunities for further improvement that we may
want to triage:
- Add examples and pictures of more realistic collected distributed
trace information to help developers internalize why it might be useful
to them.
- Add a diagnostic guide showing how distributed tracing assists in
resolving a realistic production issue.
- Ensure the getting started guides for collection transfer smoothly to
remote pages in the AppInsights and OpenTelemetry docs
  - There was a spot I wanted to link to supported OpenTelemetry
     exporters but there is no anchor that goes solely to that
     information.
- Flesh out the instrumentation guidance to answer more questions
developers may encounter such as:
  - naming conventions for Activity
  - when to use DisplayName
  - how to log exceptions
  - how to use links for batch processing systems
  - how to propagate IDs to and from network protocols
  - how to migrate pre-existing Activity instrumentation to use
     ActivitySource
- Add performance analysis and guidance to reassure developers that
  Activity instrumentation + OpenTelemetry are serious about high
  performance workloads.
- Discuss more conceptual topics and/or go into greater depth about
some areas:
  - Hierarchical vs. W3C ID.
  - Format of a W3C ID
  - Describe sampling options in ActivityListener
  - Explain why some activities are uncapturable with ActivityListener
     and require workarounds with DiagnosticListener.
 - Roadmap for handling issues around custom propagators, supporting
protocol updates, and suppressing undesired automatic Activity creation

noahfalk commented:
Please take a look @sywhang @shirhatti @tarekgh @cijothomas @reyang

noahfalk commented:
Is there a way to see a preview of what this change would look like on docs.microsoft.com without having merged it?

tarekgh commented:
@carlossanlop do you know the answer of @noahfalk question `Is there a way to see a preview of what this change would look like on docs.microsoft.com without having merged it?`?

tarekgh commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
```diff
@@ -0,0 +1,326 @@
+
+# Collect a distributed trace
+
+**This article applies to: ✔️** .NET Core 5.0 and later versions **or** any .NET application using the 
```
>✔️ [](start = 27, length = 2)

why you have this symbol?

sywhang replied:
All the diagnostics docs have this symbol for marking applicable .NET versions

reyang commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
```diff
@@ -0,0 +1,326 @@
+# Collect a distributed trace
+
+**This article applies to: ✔️** .NET Core 5.0 and later versions **or** any .NET application using the 
+[DiagnosticSource NuGet package](https://www.nuget.org/packages/System.Diagnostics.DiagnosticSource/5.0.1) version 5 or later
```
```suggestion
[DiagnosticSource NuGet package](https://www.nuget.org/packages/System.Diagnostics.DiagnosticSource/) version 5 or later
```

reyang commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
```diff
@@ -0,0 +1,326 @@
+
+# Collect a distributed trace
+
+**This article applies to: ✔️** .NET Core 5.0 and later versions **or** any .NET application using the 
```
```suggestion
**This article applies to: ✔️** .NET 5.0 and later versions **or** any .NET application using the 
```

reyang commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
```diff
@@ -0,0 +1,326 @@
+**This article applies to: ✔️** .NET Core 5.0 and later versions **or** any .NET application using the 
+[DiagnosticSource NuGet package](https://www.nuget.org/packages/System.Diagnostics.DiagnosticSource/5.0.1) version 5 or later
+
+.NET applications can be instrumented using the <xref:System.Diagnostics.Activity?displayProperty=nameWithType> API to produce
```
Is the term `application` here referring to any application and library?

noahfalk replied:
Yes that was intent, but I changed 'application' -> 'code' to avoid any implication that this is only for executables.

reyang commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
````diff
@@ -0,0 +1,326 @@
+```C#
+using OpenTelemetry;
+using OpenTelemetry.Trace;
+using System;
````
minor: put System before OpenTelemetry?

noahfalk replied:
By default the IDE sorts the usings in alphabetical order so OpenTelemetry goes first. I figured it was better to use the default ordering than to customize it.

reyang commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
```diff
@@ -0,0 +1,326 @@
+
+Setting <xref:System.Diagnostics.Activity.DefaultIdFormat?displayProperty=nameWithType> is optional
+but helps ensure the sample produces similar output on different .NET runtime versions. .NET 5.0 uses
+the W3C ID format by default but earlier .NET versions default to using 
```
W3C -> W3C TraceContext

reyang commented on docs/core/diagnostics/distributed-tracing-concepts.md:
```diff
@@ -0,0 +1,137 @@
+
+### Traces and Activities
+
+Each time a new request is received by an application a new trace can be initiated. In
```
the application can join the existing trace (if there is one from the incoming request) or a new trace can be initiated

noahfalk replied:
I reworded to this to say it can be associated with a trace without explicitly saying the trace is new. I worry that the reader doesn't have enough context yet to understand what it means to join an existing trace or that trace information flows with the request.

reyang commented on docs/core/diagnostics/distributed-tracing-concepts.md:
```diff
@@ -0,0 +1,137 @@
+In order to track work across process boundaries Activity parent IDs need to be transmitted across
+the network so that the receiving process can create Activities that refer to them. When using
+W3C standard Activity IDs (<xref:System.Diagnostics.ActivityIdFormat.W3C%2A?displayProperty=nameWithType>)
+.NET will also use [W3C standard networking protocols](https://www.w3.org/TR/trace-context/) to
```
```suggestion
.NET will also use [W3C TraceContext](https://www.w3.org/TR/trace-context/) to
```

shirhatti commented:
> Is there a way to see a preview of what this change would look like on docs.microsoft.com without having merged it?

https://review.docs.microsoft.com/en-us/dotnet/core/diagnostics/distributed-tracing?branch=pr-en-us-23313

tarekgh commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
```diff
@@ -0,0 +1,326 @@
+            ActivitySource.AddActivityListener(new ActivityListener()
+            {
+                ShouldListenTo = (source) => true,
+                Sample = (ref ActivityCreationOptions<ActivityContext> options) => ActivitySamplingResult.AllDataAndRecorded,
```
>Sample [](start = 16, length = 6)

did you try running this sample on Full Framework?

noahfalk replied:
Thanks, that did catch an error, there was an errant using System.Net.Http; left over that is harmless of Core but fails to build on net48.

reyang commented on docs/core/diagnostics/distributed-tracing-instrumentation-walkthroughs.md:
```diff
@@ -0,0 +1,412 @@
+        {
+            using (Activity activity = source.StartActivity("SomeWork"))
+            {
+                activity?.SetTag("foo", foo);
```
Do we want to encourage people to set simple tags in the `StartActivity` call (to avoid extra API invocations) and only use `activity?.SetTag` for tags that are expensive to evaluate?

noahfalk replied:
I'm less worried about API invocations as I am allocations. Specifying tags in StartActivity requires allocating an IEnumerable and the allocation still happens even for unsampled Activities. I'm glad to have us change the guidance if we decide specifying them in StartActivity() is preferable, but so far all examples I recall (including the [OpenTelemetry ones](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/docs/trace/getting-started/Program.cs#L36)) use the SetTag() pattern.

cijothomas replied:
Otel spec recommendation:
"Whenever possible, users SHOULD set any already known attributes at span creation instead of calling SetAttribute later."
https://github.com/open-telemetry/opentelemetry-specification/blob/v1.0.1/specification/trace/api.md#span-creation

But agree, its *not* free in .NET as it requires alloc before knowing sampling decision.

noahfalk replied:
Do you have any background on that spec recommendation? I could guess it is designed to make it more reliable for samplers to sample using attributes but I've never seen an example of any sampler that cared about attributes. It makes me sad to trade away performance in cases that apps aren't even logging telemetry for a scenario that appears to be rare?

cijothomas replied:
I have seen people asking about this and some writing custom sampler (which wraps the default ones) to sample out certain paths based on known attributes.

(Btw, I try to interpret it as - if the attributes are readily available (i.e available already as IEnumerable for whatever other reasons - like metric/log purpose, then use it in Activity creation, since this should be relatively free.)

noahfalk replied:
yeah if the attributes are already in an IEnumerable that would be great, though I am suspicious that is going to be the rare case.

shirhatti commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
````diff
@@ -0,0 +1,326 @@
+.NET versions add the [System.Diagnostics.DiagnosticSource NuGet package](https://www.nuget.org/packages/System.Diagnostics.DiagnosticSource/)
+version 5.0.1 or greater.
+```dotnetcli
+dotnet add package System.Diagnostics.DiagnosticSource --version 5.0.1
````
```suggestion
dotnet add package System.Diagnostics.DiagnosticSource
```

I'd rather not have versions burned into the doc since they tend to go stale

noahfalk replied:
For some reason I had concluded that was important, but retesting without it worked fine so let's say it was user error : )

tarekgh commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
````diff
@@ -0,0 +1,326 @@
+Example work done
+```
+
+Setting <xref:System.Diagnostics.Activity.DefaultIdFormat?displayProperty=nameWithType> is optional
````
>DefaultIdFormat [](start = 42, length = 15)

Should mention ForceDefaultIdFormat property too? 

I would suggest adding the usage of these properties to the sample with nice comment there.

noahfalk replied:
I added ForceDefaultIdFormat here and more discussion of both properties to the Conceptual doc which is linked.

shirhatti commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
```diff
@@ -0,0 +1,326 @@
+
+        static async Task StepOne()
+        {
+            using (Activity a = s_source.StartActivity("StepOne"))
```
Do we want some explicit guidance on what meets the bar for Activity creation? Creating Activities is expensive and I don't this to be misunderstood as create an Activity for every method.

noahfalk replied:
There is a little bit of guidance in the instrumentation tutorial. Given this is the collection tutorial we'd be wandering a bit off topic to discuss best practices for instrumentation.

sywhang commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
```diff
@@ -0,0 +1,326 @@
+---
+title: Collect a distributed trace - .NET
+description: A tutorial to collect distributed traces in .NET applications
```
We may want to add `OpenTelemetry` in the description to get some SEO boost.

shirhatti commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
```diff
@@ -0,0 +1,326 @@
+        public static async Task Main()
+        {
+            using var tracerProvider = Sdk.CreateTracerProviderBuilder()
+                .SetSampler(new AlwaysOnSampler())
```
```suggestion
               .SetResourceBuilder(ResourceBuilder.CreateDefault().AddService("MySample"))
                .SetSampler(new AlwaysOnSampler())
```

cijothomas replied:
To keep this a minimal code, I'd suggest to remove both Sampler and Resource concept.

noahfalk replied:
I have removed both

noahfalk replied:
I've added Resource

shirhatti commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
```diff
@@ -0,0 +1,326 @@
+Activity.StartTime:   2021-03-15T01:58:10.7661575Z
+Activity.Duration:    00:00:00.5013965
+Resource associated with Activity:
+    service.name: unknown_service:temp
```
Update service name as mentioned in comment above

noahfalk replied:
@cijothomas recommend removing the API call so we have to decide, do we want simpler code or do we want this field populated.

I'm leaning toward including it in the example because it seems like something a typical OpenTelemetry user will need to do and understand. Thoughts?

cijothomas replied:
If the example should look neater, then yes, need to add Resource in the code example.

shirhatti commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
```diff
@@ -0,0 +1,326 @@
+> specific API for adding custom distributed tracing information.
+
+
+## Collect using a custom logging implementation
```
Could we not use the word logging here? I was confused for a second on why we're talking about logging here 😄

noahfalk replied:
Reworded : )

shirhatti commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
````diff
@@ -0,0 +1,326 @@
+Example work done
+```
+
+Setting <xref:System.Diagnostics.Activity.DefaultIdFormat?displayProperty=nameWithType> is optional
````
nit: This xref is broken

shirhatti commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
```diff
@@ -0,0 +1,326 @@
+Activity is associated with an ActivitySource which acts as a namespace for a set of Activities.
+This callback is invoked once for each ActivitySource in the process. Returning true indicates
+the listener should be notified about Activities associated with this source.
+ - <xref:System.Diagnostics.ActivityListener.Sample?displayProperty=nameWithType> - By default
```
nit: Broken xref <xref:System.Diagnostics.ActivitySource.StartActivity?displayProperty=nameWithType>

tarekgh commented on docs/core/diagnostics/distributed-tracing-instrumentation-walkthroughs.md:
```diff
@@ -0,0 +1,412 @@
+
+# Adding distrubuted tracing instrumentation
+
+**This article applies to: ✔️** .NET Core 5.0 and later versions **or** any .NET application using the 
```
>✔️ [](start = 27, length = 2)

ditto

sywhang reviewed (commented):
@noahfalk Can you also add the new pages you added as links in docs/core/diagnostics/logging-tracing.md, and docs/fundamentals/toc.yml?

tarekgh commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
```diff
@@ -0,0 +1,326 @@
+using System.Net.Http;
+using System.Threading.Tasks;
+
+namespace Sample.DistributedTracing
```
>DistributedTracing [](start = 17, length = 18)

This example is good but not using the created activities at all which can be confusing to the reader why we are creating activity at all.

noahfalk replied:
Did you have a suggestion for what to change? The text at the top currently says "First you will create an example application that has some distributed trace instrumentation but nothing is being collected"

cijothomas replied:
We can leave a code comment "In this example, the activity will not be created at all, as there is no registered listener."

noahfalk replied:
I updated the text above the code snippet to make it hopefully more explicit : )

sywhang commented on docs/core/diagnostics/distributed-tracing-instrumentation-walkthroughs.md:
````diff
@@ -0,0 +1,412 @@
+```csharp
+        static async Task DoSomeWork(string foo, int bar)
+        {
+            using (Activity activity = source.StartActivity("SomeWork"))
````
Can we change this to Activity? so that it's more obvious this can return null?

```suggestion
            using (Activity? activity = source.StartActivity("SomeWork"))
```

noahfalk replied:
We can only use "Activity?" if we are using C# 8 which isn't the default version for a new console app in the 3.1 SDK. I assume we could force it forward but I worry it will generate compilation errors when people copy/paste the pattern into other projects. 

I'm planning to leave this as-is but leaving the issue open if others think we should switch it?

tarekgh commented on docs/core/diagnostics/distributed-tracing-instrumentation-walkthroughs.md:
```diff
@@ -0,0 +1,412 @@
+---
+title: Add distributed tracing instrumentation - .NET
```
>title [](start = 0, length = 5)

as we removed some of the contents from the distributed-tracing doc, I think we need to cover some more information in this doc (e.g. ActivityLinks and ActivityKind).

noahfalk replied:
Added

tarekgh commented:
Will be good if track deleting the following files after merging this change 
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Diagnostics.DiagnosticSource/src/ActivityUserGuide.md
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Diagnostics.DiagnosticSource/src/DiagnosticSourceUsersGuide.md

tarekgh reviewed (approved):
Great writing! 
modulo some minor comments I added and the other reviewers' comments, LGTM.

gewarren commented:
> Is there a way to see a preview of what this change would look like on docs.microsoft.com without having merged it?

Here's a preview: https://review.docs.microsoft.com/en-us/dotnet/core/diagnostics/distributed-tracing-collection-walkthroughs?branch=pr-en-us-23313. You can access the preview URLs from the [build report](https://opbuilduserstoragepublic.blob.core.windows.net/report/2021%5C3%5C15%5Ca3bda507-3390-de91-8f7f-26f90f4e5fc8%5CPullRequest%5C202103151247178852-23313%5Cworkflow_report.html?sv=2016-05-31&sr=b&sig=CdGgXKIDAZsJSaq97VKeppUEtK1HrECzsibiBi%2FrTLs%3D&st=2021-03-15T12%3A45%3A04Z&se=2021-04-15T12%3A50%3A04Z&sp=r), which you access by clicking "Details" next to the OpenPublishing.Build check. I noticed this articles isn't part of the table of contents, but it should be. Also, there are a number of build warnings to address (bad links).

gewarren commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
```diff
@@ -0,0 +1,326 @@
+This callback is invoked once for each ActivitySource in the process. Returning true indicates
+the listener should be notified about Activities associated with this source.
+ - <xref:System.Diagnostics.ActivityListener.Sample?displayProperty=nameWithType> - By default
+<xref:System.Diagnostics.ActivitySource.StartActivity?displayProperty=nameWithType> does not
```
```suggestion
<xref:System.Diagnostics.ActivitySource.StartActivity%2A?displayProperty=nameWithType> does not
```

This links to the overloads list for StartActivity. Is that what you want?

noahfalk replied:
I think the overload list would be fine, the statement is accurate for all overloads of the method. Thanks!

gewarren commented on docs/core/diagnostics/distributed-tracing-concepts.md:
```diff
@@ -0,0 +1,137 @@
+
+In order to track work across process boundaries Activity parent IDs need to be transmitted across
+the network so that the receiving process can create Activities that refer to them. When using
+W3C standard Activity IDs (<xref:System.Diagnostics.ActivityIdFormat.W3C%2A?displayProperty=nameWithType>)
```
```suggestion
W3C standard Activity IDs (<xref:System.Diagnostics.ActivityIdFormat.W3C?displayProperty=nameWithType>)
```

gewarren commented on docs/core/diagnostics/distributed-tracing-concepts.md:
```diff
@@ -0,0 +1,137 @@
+W3C standard Activity IDs (<xref:System.Diagnostics.ActivityIdFormat.W3C%2A?displayProperty=nameWithType>)
+.NET will also use [W3C standard networking protocols](https://www.w3.org/TR/trace-context/) to
+transmit this information. When using the 
+<xref:System.Diagnostics.ActivityIdFormat.Hierarchical%2A?displayProperty=nameWithType> ID format
```
```suggestion
<xref:System.Diagnostics.ActivityIdFormat.Hierarchical?displayProperty=nameWithType> ID format
```

gewarren commented on docs/core/diagnostics/distributed-tracing-concepts.md:
```diff
@@ -0,0 +1,137 @@
+information necessary to propagate distributing tracing IDs, or to populate it with complete
+diagnostic information. These choices trade-off increasing performance overhead for
+increasing diagnostic utility. Activities that are started using the older pattern of invoking
+<xref:System.Diagnostics.Activity..ctor%2A?displayProperty=nameWithType> and 
```
```suggestion
<xref:System.Diagnostics.Activity.%23ctor%2A?displayProperty=nameWithType> and 
```

gewarren commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
```diff
@@ -0,0 +1,326 @@
+stops.
+
+Once an ActivityListener has been created and the callbacks are populated, invoking
+<xref:System.Diagnostics.ActivitySource.AddActivityListener?displayProperty=nameWithType>
```
```suggestion
<xref:System.Diagnostics.ActivitySource.AddActivityListener(System.Diagnostics.ActivityListener)?displayProperty=nameWithType>
```

gewarren reviewed (commented):
Fixed some xrefs.

cijothomas commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
```diff
@@ -0,0 +1,326 @@
+
+# Collect a distributed trace
+
+**This article applies to: ✔️** .NET Core 5.0 and later versions **or** any .NET application using the 
```
minor:
I'd try to word this to say that - This is applicable for all .NET versions, and then further clarify that - .NET 5 or higher, requires no extra dependency, older versions require installing SD.DS package.
This will avoid any potential misunderstanding that this is only a new thing and only works for newer .NET only.
(#1 question we received in otel as well!)

noahfalk replied:
Good feedback. I've update the 'applies to' guidance to indicate .NET Framework >= 4.5 and .Net Core >= 2.1. The need to include DiagnosticSource NuGet is embedded in the tutorial text but no longer mentioned here.

cijothomas commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
```diff
@@ -0,0 +1,326 @@
+[DiagnosticSource NuGet package](https://www.nuget.org/packages/System.Diagnostics.DiagnosticSource/5.0.1) version 5 or later
+
+.NET applications can be instrumented using the <xref:System.Diagnostics.Activity?displayProperty=nameWithType> API to produce
+distributed tracing telemetry. In this tutorial you will record this instrumented telemetry with different telemetry collection
```
record this instrumented telemetry with different telemetry collection libraries -- this sentence is bit hard to follow.

noahfalk replied:
yep : ) Updated the wording.

cijothomas commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
```diff
@@ -0,0 +1,326 @@
+
+### Prerequisites
+
+- [.NET Core 3.1 SDK](https://dotnet.microsoft.com/download/dotnet) or a later version
```
Why 3.1 as the pre-req? This would work even in 2.1 right?

noahfalk replied:
Probably, I just didn't test it. I'm going to leave it as-is for now but if customers care we could re-test the workflow using an older SDK to ensure I didn't accidentally use a 3.1 only feature.

cijothomas commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
````diff
@@ -0,0 +1,326 @@
+Example work done
+```
+
+### Add logging using OpenTelemetry
````
I'd avoid the work "logging" here as it could be confused with the "logging" pillar of Observability. Would prefer to consistently use `tracing` in this doc.

noahfalk replied:
I reworded as "Collect a trace using OpenTelemetry".

I am guessing that customers will relate better to 'trace' or 'tracing data' as nouns rather than 'trace/tracing' as a verb. I'll try to be consistent using the word 'collect' when I am drawing attention to the work of capturing data and transmitting it to a backend.

cijothomas commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
```diff
@@ -0,0 +1,326 @@
+            using var tracerProvider = Sdk.CreateTracerProviderBuilder()
+                .SetSampler(new AlwaysOnSampler())
+                // Add more libraries
+                .AddSource("Sample.DistributedTracing")
```
It'd be good to clarify that this name must match the name of AS created earlier.

noahfalk replied:
Updated the text after the example to clarify this

cijothomas commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
```diff
@@ -0,0 +1,326 @@
+
+The console exporter is helpful for quick examples or local development but in a production deployment
+you will probably want to send logs to a central logging store. OpenTelemetry supports a variety
+of different logging destinations using different
```
logging - I'll try to avoid this term :)

noahfalk replied:
Reworded : )

cijothomas commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
```diff
@@ -0,0 +1,326 @@
+## Collect using Application Insights
+
+Distributed tracing telemetry is automatically captured after configuring the Application Insights SDK 
+([.NET](https://docs.microsoft.com/en-us/azure/azure-monitor/app/asp-net), [.NET Core](https://docs.microsoft.com/en-us/azure/azure-monitor/app/asp-net-core))
```
```suggestion
([.NET](https://docs.microsoft.com/en-us/azure/azure-monitor/app/asp-net), [.NET Core](https://docs.microsoft.com/azure/azure-monitor/app/asp-net-core))
```

cijothomas replied:
Need to do this for all doc links.

cijothomas commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
```diff
@@ -0,0 +1,326 @@
+## Collect using Application Insights
+
+Distributed tracing telemetry is automatically captured after configuring the Application Insights SDK 
+([.NET](https://docs.microsoft.com/en-us/azure/azure-monitor/app/asp-net), [.NET Core](https://docs.microsoft.com/en-us/azure/azure-monitor/app/asp-net-core))
```
```suggestion
([ASP.NET](https://docs.microsoft.com/en-us/azure/azure-monitor/app/asp-net), [ASP.NET Core](https://docs.microsoft.com/en-us/azure/azure-monitor/app/asp-net-core))
```

cijothomas commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
```diff
@@ -0,0 +1,326 @@
+ - <xref:System.Diagnostics.ActivityListener.ActivityStarted?displayProperty=nameWithType> and
+<xref:System.Diagnostics.ActivityListener.ActivityStopped?displayProperty=nameWithType> are
+called when an Activity is started and stopped respectively. These callbacks provide an
+oportunity to log any relevant information about the Activity. When an Activity has just
```
```suggestion
oportunity to add more information (Tags, Events, etc.) to the Activity. When an Activity has just
```

noahfalk replied:
Reworded a bit. I want to keep the emphasis on reading from the Activity but I did mention that it is possible to modify it as well.

cijothomas commented:
> Will be good if track deleting the following files after merging this change
> https://github.com/dotnet/runtime/blob/main/src/libraries/System.Diagnostics.DiagnosticSource/src/ActivityUserGuide.md
> https://github.com/dotnet/runtime/blob/main/src/libraries/System.Diagnostics.DiagnosticSource/src/DiagnosticSourceUsersGuide.md

I'll still keep them both and add ta section at the beginning to point to new guidance.
I find the DiagnosticSourceUserGuide to be relevant (to publish strong objects as payloads) even without Activity.

noahfalk commented on docs/core/diagnostics/distributed-tracing-instrumentation-walkthroughs.md:
````diff
@@ -0,0 +1,412 @@
+some events to the Activity:
+
+```csharp
+        static async Task DoSomeWork(string foo, int bar)
````
The indent aligns the code properly for me when it is copy-pasted into the example code given above. Were you attempt to paste it or was this just for the visual aesthetics? 

If it is only for visuals I'd suggest we leave it as-is for easy copy/pasting, however if it is also messing up copy/paste then I'd want to change it.

noahfalk committed:
Applying review feedback

carlossanlop commented:
> @carlossanlop Carlos Sanchez Lopez FTE do you know the answer of @noahfalk Noah Falk FTE question `Is there a way to see a preview of what this change would look like on docs.microsoft.com without having merged it?`?

@tarekgh Seems to be similar to the one in dotnet-api-docs:

In the CI section, click on "Show all checks" to expand the CI legs. There's one called `OpenPublishing.Build`. On the right side, click on "Details", which will [open a page](https://opbuilduserstoragepublic.blob.core.windows.net/report/2021%5C3%5C17%5Ca3bda507-3390-de91-8f7f-26f90f4e5fc8%5CPullRequest%5C202103171340336011-23313%5Cworkflow_report.html?sv=2016-05-31&sr=b&sig=ICOfQB58IUGx4J8fmm2pVEFevDzMdwzBO2G3TNXH3Wk%3D&st=2021-03-17T13%3A38%3A19Z&se=2021-04-17T13%3A43%3A19Z&sp=r) with some details about the build.

The "Validated files" table (first one) is the one you care about: click on Preview for each link. Or, click on just one, and look at the preview of your new pages, and navigate to the other pages from that one (the links should work).

By the way, the `OpenPublishing.Build` leg has a warning, which you need to fix: there is a cross reference to a DocId that is not correct. The full list of warnings can be consulted in the details page, but you only care about those coming from files you added in this PR. All others can be ignored (The Docs team should take care of those).

Hope that helped!

noahfalk committed:
Adding OTel resource

noahfalk commented:
> @noahfalk Can you also add the new pages you added as links in docs/core/diagnostics/logging-tracing.md, and docs/fundamentals/toc.yml?

I've added entries to the toc and updated the text in logging-tracing.md but stopped short of adding all the links to logging-tracing.md. It wasn't clear that there was much value in duplicating all the links from distributed-tracing.md into logging-tracing.md if that means we now have to keep both pages in sync and they are replicas of each other. I am guessing that anyone interested in distributed tracing will click through from logging-tracing.md -> distributed-tracing.md -> the specific page they wanted. If you see any sign that people aren't finding the info they want because links aren't there I'm glad to change it.

noahfalk committed:
Updating toc

noahfalk commented:
> > Will be good if track deleting the following files after merging this change
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Diagnostics.DiagnosticSource/src/ActivityUserGuide.md
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Diagnostics.DiagnosticSource/src/DiagnosticSourceUsersGuide.md

> I'll still keep them both and add ta section at the beginning to point to new guidance.
I find the DiagnosticSourceUserGuide to be relevant (to publish strong objects as payloads) even without Activity.

Agreed. DiagnosticSource is mostly not about distributed tracing so I only updated the Activity Guide:
https://github.com/dotnet/runtime/pull/49818

noahfalk committed:
Fix xref

noahfalk commented:
Thanks for all the great feedback all! I've addressed almost everything but there are a few that it is unclear how to best address. Let me know if you are happy with the current state of the PR or you think we need to make more changes before checking in. If I don't hear anything today I'll assume it is OK to merge and we can address lingering issues with specific follow up PRs. Thanks!

cijothomas commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
```diff
@@ -0,0 +1,365 @@
+
+### Prerequisites
+
+- [.NET Core 3.1 SDK](https://dotnet.microsoft.com/download/dotnet) or a later version
```
nit: to avoid the misconception that "Collect traces using OpenTelemetry" *requires* .NET Core 3.1, I'd suggest to keep this .NET Core 2.1

cijothomas commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
```diff
@@ -0,0 +1,365 @@
+[OpenTelemetry getting started guide](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/docs/trace/getting-started/README.md)
+for more information.
+
+Add the [OpenTelemetry](https://www.nuget.org/packages/OpenTelemetry/) and
```
Installing `OpenTelemetry.Exporter.Console` automatically brings `OpenTelemetry`, so this step may be implict.

cijothomas commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
````diff
@@ -0,0 +1,365 @@
+}
+```
+
+Running the app does not record any tracing information yet
````
```suggestion
Running the app does not collect any trace data yet:
```

cijothomas commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
````diff
@@ -0,0 +1,365 @@
+Example work done
+```
+
+### Add code to observe the Activities
````
```suggestion
### Add code to collect the traces
```

cijothomas commented on docs/core/diagnostics/distributed-tracing-collection-walkthroughs.md:
```diff
@@ -0,0 +1,365 @@
+- <xref:System.Diagnostics.ActivityListener.ShouldListenTo> - Each
+Activity is associated with an ActivitySource which acts as a namespace for a set of Activities.
+This callback is invoked once for each ActivitySource in the process. Returning true indicates
+the listener should be notified about Activities associated with this source.
```
you may add something to the tune of "Returning false means the `Sample` callback is not invoked at all.."

noahfalk replied:
I updated the text for 'true' path to make it clear you needed to pick that if you wanted to do sampling. Hopefully that is explicit enough but we can always make further edits as needed.

cijothomas commented on docs/core/diagnostics/distributed-tracing-concepts.md:
```diff
@@ -0,0 +1,137 @@
+[the standard](https://www.w3.org/TR/trace-context/) to transmit this information. When using the
+<xref:System.Diagnostics.ActivityIdFormat.Hierarchical> ID format
+.NET uses a custom request-id HTTP header to transmit the ID. Unlike many other language runtimes
+.NET in-box libraries natively understand how to decode and encode Activity IDs on HTTP messages
```
might be a good idea to clarify what is "in-box" libraries. You are referring to ASP.NET Core and HttpClient in .NET Core right?

cijothomas commented on docs/core/diagnostics/distributed-tracing-concepts.md:
```diff
@@ -0,0 +1,137 @@
+DiagnosticListener is a more complex scenario that requires the instrumented code to opt-in by
+invoking <xref:System.Diagnostics.DiagnosticSource.StartActivity%2A?displayProperty=nameWithType> and
+the collection library needs to know the exact naming information that the instrumented code
+used when starting it. Using DiagnosticSource and DiagnosticListener allows the creator
```
We could state that the in-box libraries publish DiagnosticSource events to pass things like the HttpContext instances, which may be leveraged by listeners to extract additional information to enrich the Activity.

noahfalk replied:
I think I will avoid going to into that level of detail here. I agree there is lots more we could say about DiagnosticSource which could probably could fill its own pages.

cijothomas commented on docs/core/diagnostics/distributed-tracing-concepts.md:
```diff
@@ -0,0 +1,137 @@
+### Sampling
+
+For improved performance in high throughput applications, distributed tracing on .NET supports
+sampling only a subset of requests rather than recording all of them. For activites created with
```
```suggestion
sampling only a subset of traces rather than recording all of them. For activites created with
```

cijothomas commented on docs/core/diagnostics/distributed-tracing-concepts.md:
```diff
@@ -0,0 +1,137 @@
+For improved performance in high throughput applications, distributed tracing on .NET supports
+sampling only a subset of requests rather than recording all of them. For activites created with
+the recommended <xref:System.Diagnostics.ActivitySource.StartActivity%2A?displayProperty=nameWithType>
+API, logging libraries can control sampling with the
```
```suggestion
API, telemetry collection libraries can control sampling with the
```
prefer to to avoid "logging"

noahfalk replied:
arg, thanks, I missed this one : ) I had been trying to change it everywhere

cijothomas commented on docs/core/diagnostics/distributed-tracing-instrumentation-walkthroughs.md:
```diff
@@ -0,0 +1,456 @@
+- The version parameter is optional. It is recommended to provide the version in case you release multiple
+versions of the library and make changes to the instrumented telemetry.
+
+> [!NOTE]
```
Good writing!

cijothomas commented on docs/core/diagnostics/distributed-tracing-instrumentation-walkthroughs.md:
```diff
@@ -0,0 +1,456 @@
+stores ID information about the Activity being linked to. ActivityContext can be retrieved from in-process
+Activity objects using <xref:System.Diagnostics.Activity.Context%2A?displayProperty=nameWithType> or
+it can be parsed from serialized id information using
+<xref:System.Diagnostics.ActivityContext.Parse(System.String,System.String)?displayProperty=nameWithType>.
```
add a note saying - ActivityLinks can only be provided at the source.StartActivityTime, and cannot be added later.
(After reading AddActivityEvent API above, its can be mistakenly assumed that links can also be added later..)

sywhang reviewed (approved):
LGTM - if there is no other suggestions by EOD I'll go ahead and merge this. Thanks everyone!

cijothomas commented on docs/core/diagnostics/distributed-tracing-instrumentation-walkthroughs.md:
```diff
@@ -0,0 +1,456 @@
+    using(Activity activity = s_source.StartActivity(name: "BigBatchOfWork",
+                                                     kind: ActivityKind.Internal,
+                                                     parentContext: null,
+                                                     links: requestIds.Select(ctx => new ActivityLink(ctx))
```
`requestIds` is not defined anywhere?

noahfalk replied:
whoops, meant to be requestContexts, thanks!

cijothomas commented on docs/core/diagnostics/distributed-tracing-instrumentation-walkthroughs.md:
```diff
@@ -0,0 +1,456 @@
+
+## Optional: Adding Events
+
+Events are timestamped messages that can attach an arbitrary stream of additional diagnostic data to Activities. Add
```
link to OpenTelemetry "Event" https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#add-events

noahfalk replied:
I've thought about it and concluded I'd rather not lean too heavily on the OpenTelemetry specification at this point. I'm trying to strike a balance - on the one hand I want .NET's Activity instrumentation to stand on its own without needing an external definition, on the other I want developers to understand they can use OpenTelemetry to collect it. I am betting that developers familiar with OTel's model will find the mapping intuitive. For developer's not familiar with OTel's model I don't want them to feel that they have to understand OTel's spec to understand the API.

cijothomas commented on docs/core/diagnostics/distributed-tracing-instrumentation-walkthroughs.md:
```diff
@@ -0,0 +1,456 @@
+
+When work occurs in batch processing systems a single Activity might represent work on behalf of many
+different requests simultaneously, each of which has its own trace-id. Although Activity is restricted
+to have a single parent, it can link to additional trace-ids using
```
Link to OTel notion of links https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/overview.md#links-between-spans

noahfalk replied:
Similar to Events, I'm going to avoid linking to the OTel spec here under the premise the .NET definitions of these concepts needs to stand on its own.

cijothomas reviewed (approved):
Excellent write-up!
Left few small comments.
Only strong recommendation is to avoid using the "logging" term in tracing doc.

noahfalk committed:
Addressing more feedback

noahfalk committed:
More feedback

noahfalk committed:
newline

noahfalk committed:
last touchups I hope
